RFC-0005-functions-stats.md

```diff
@@ -0,0 +1,561 @@
+replace `scalar_function_stats_propagation_enabled` true or false to `scalar_function_stats_propagation_accuracy` value will be in **[low, medium, high, accurate, none]**
+
+* `none` will turn off this feature completely 
+* `accurate` will propagate stats for only those functions which are marked as having accurate stats.
+* similarly `low`, `medium` and `high` can be set up. With `low` being the most permissible mode, i.e. 
+    functions with `low` as configured accuracy for stats will have their stats propagated.
+
+A few examples:
+
+1. String function `reverse(Col)` configured with `propagateAllStats=true` the accuracy confidence level is `accurate` for this function.
+2. String function `upper(Col)` configured with `propagateAllStats=true` the accuracy confidence level is `medium` for this function as NDV can vary if the input records have
+    entries with same string as upper case and lower case.
+3. An example of `low` accuracy confidence level is concat function, 
+```java
+@Description("concatenates given character strings")
+@ScalarFunction
+@LiteralParameters({"x", "y", "u"})
+@Constraint(variable = "u", expression = "x + y")
+@SqlType("char(u)")
+public static Slice concat(@LiteralParameter("x") Long x,
+        @ScalarPropagateSourceStats(
+                nullFraction = SUM_ARGUMENTS,
+                avgRowSize = SUM_ARGUMENTS,
+                distinctValuesCount = NON_NULL_ROW_COUNT) @SqlType("char(x)") Slice left,
+        @SqlType("char(y)") Slice right)
+{ /*...*/ }
+```
+where distinct values count is an estimated value assuming every non null entry is unique, which may not be always true. 
+
+#### Define an expression language:
+The following example is a more powerful form of expressing how to estimate stats for functions based on their characteristics. Here,
+we propose the use of expressions ( i.e. an expression language with a fixed grammar ) In the following example of `substr` :
+For average row size we have used the value of length argument. For distinct values count it is more complex, e.g. if length is same as
+the upper bound for `varchar(x)` i.e. `x` , then use source stat `distinct values count` of the argument `utf8`. 
+```java
+    @Description("substring of given length starting at an index")
+    @ScalarFunction
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice substr(
+            @ScalarPropagateSourceStats(
+                    nullFraction = USE_SOURCE_STATS,
+                    distinctValuesCount = "value(arg_length) <3 ? estimate_ndv(x, value(arg_length), BASIC_LATIN, min(output_row_count, source_stats_ndv())) * (1- source_stats_null_fraction()) : UNKNOWN",
+                    avgRowSize = "value(arg_length)" ) @SqlType("varchar(x)") Slice utf8,
+            @SqlType(StandardTypes.BIGINT) long start,
+            @SqlType(StandardTypes.BIGINT) long length)
+    { //...
+    }
+```
+
+#### virtual function
```

Comment:
It would be helpful to discuss changes required in Velox with an issue or discussion in the Velox repo.